docs/internals.md

```diff
@@ -0,0 +1,235 @@
+# mquery internals
+
+## How does it work?
+
+YARA is pretty fast, but searching through large dataset for given
+signature can take a lot of time. We pre-filter the results, so it is only
+necessary to run YARA against a small fraction of binaries:
+
+![mquery flowchart](docs/mquery-flowchart.png?raw=1)
+
+Internally this is implemented with a n-gram database called
+[ursadb](https://github.com/CERT-Polska/ursadb/).
+See [readme](https://github.com/CERT-Polska/ursadb/) for more details.
+
+In short, we look for short (3-4) byte fragments in indexed files, and we can
+tell (almost) immediately which files contain given 3-byte pattern. So for
+example if you have files:
+
+```
+echo "abcde" > abcde
+echo "abcxy" > abcxy
+echo "asdfg" > xbcdx
+```
+
+And rule:
+
+```
+rule example
+{
+    strings:
+        $test = "abcd"
+    condition:
+        $test
+}
+```
+
+We parse it internally to `"abc" AND "bcd"`, execute both parts of the query
+separately and merge the result.
+
+# Known limitations and design decisions.
+
+Mquery's goal is to accelerate Yara queries. It should **always** return the
+same results as running Yara on the dataset naively. If it doesn't, please
+report a bug.
+
+Because of the specifics of the database engine, we only accelerate a pretty
+naive subset of Yara. Additionaly false positives are possible during the
+filtering stage. So we still have to re-check all the potential
+results with the real Yara binary.
+
+By the way, for parsing yara rules we use and highly recommend
+[yaramod](https://github.com/avast/yaramod) library by Avast.
+
+To give you a rough idea what works and what didn't, examples of the
+accelerator limitations:
+
+1. Counting strings:
+
+```
+rule CountExample
+{
+    strings:
+        $a = "dummy1"
+        $b = "dummy2"
+
+    condition:
+        #a == 6 and #b > 10
+}
+```
+
+Is parsed to `("dummy1" AND "dummy2")`. Counting occurences is done by Yara on
+the later stage.
+
+2. `at` construct:
+
+```
+rule AtExample
+{
+    strings:
+        $a = "dummy1"
+        $b = "dummy2"
+
+    condition:
+        $a at 100 and $b at 200
+}
+```
+
+Is parsed to `("dummy1" AND "dummy2")`. Verifying the location is done by Yara
+on the later stage.
+
+3. `in` construct:
+
+```
+rule InExample
+{
+    strings:
+        $a = "dummy1"
+        $b = "dummy2"
+
+    condition:
+        $a in (0..100) and $b in (100..filesize)
+}
+```
+
+Is parsed to `("dummy1" AND "dummy2")`. Again, further verification
+will be done in a second stage.
+
+
+4. Variables:
+
+```
+rule FileSizeExample
+{
+    condition:
+       filesize > 200KB
+}
+```
+
+This is parsed to `()`. We don't accelerate file size queries, so
+*all the files* will have to be scanned with yara rule.
+
+Similarly we can't speed up other expressions like:
+
+```
+rule IssssPE
```

Comment:
digression: we *could* actually think about ripping `{4d 5a} AND {50 4d 00 00}` from that one 🤔

Reply:
Yeah, but 0x5A4D is a bit pointless (will get ignored anyway), and 0x00004550 is...tricky.

For example, what about:
```
rule IsPE
{
  condition:
     uint32(0x3C) + 1 == 0x00004550
}
```

Needs further consideration I guess.